### PR TITLE
Optimize Smart Contract to reduce gas fees

### PR DESCRIPTION
**Pull request by Arpaninformationtech**
Fixes: #7
Description: Optimization of smart contract to reduce gas fees:

I have incorporated the changes as said by @J0SAL also incorporated the changes proposed by @shikhar229169
1. The id field in the Transaction struct has been changed to uint256 type to allocate sequential IDs.
2. The Tracker struct now includes a counter member to keep track of the current ID to allocate.
3. The addExpense and addIncome functions no longer take the _id parameter from the user. Instead, they allocate sequential IDs using the counter member in the Tracker struct.
4. Added checks for the category and date parameters in the addExpense and addIncome functions to ensure they are not empty.
5. The tracker mapping now maps transaction IDs (uint256) to the corresponding string representation ("TX-X") to keep track of transactions.
6. The _getTransactionId function converts the numeric ID to a string representation with a prefix of "TX-".
7. The getUserTransactions function checks the tracker mapping to retrieve all valid transactions for the user.
8. A new uintToString function has been added to convert a uint256 ID to its string representation for the transaction ID generation.
With these changes, the contract will allocate sequential IDs for transactions and enforce checks on the category and date parameters, ensuring users cannot manipulate them through external interactions like Etherscan.

### CODE REVIEW

File: backend/contracts/ExpenseTracker.sol [hunk 1]
```diff
@@ -17,75 +18,90 @@ contract ExpenseTracker {
...
         uint256 _amount,
         string memory _description,
         string memory _date,
         string memory _category
     ) public {
-        require(all_ids[_id] == false, "invalid id");
-        transactions[msg.sender][_id] = Transaction(
-            _id,
+        require(bytes(_category).length > 0, "category is required");
+        require(bytes(_date).length > 0, "date is required");
```
Comment (shikhar229169) [inline backend/contracts/ExpenseTracker.sol:65]: Actually we must have a check that date is entered in correct format (YYYY-MM-DD), and it must not exceed current date.

File: backend/contracts/ExpenseTracker.sol [hunk 2]
```diff
@@ -17,75 +18,90 @@ contract ExpenseTracker {
...
     struct Tracker {
-        uint total;
-        uint deleted;
-        string[] transactions;
+        uint256 total;
+        uint256 deleted;
+        uint256 counter;
+        mapping(uint256 => string) transactions;
     }
-    mapping(address => mapping(string => Transaction)) private transactions;
```
Comment (shikhar229169) [inline backend/contracts/ExpenseTracker.sol:24]: Why is this deleted?

File: backend/contracts/ExpenseTracker.sol [hunk 3]
```diff
@@ -17,75 +18,90 @@ contract ExpenseTracker {
...
         string memory _date,
         string memory _category
     ) public {
-        require(all_ids[_id] == false, "invalid id");
-        transactions[msg.sender][_id] = Transaction(
-            _id,
+        require(bytes(_category).length > 0, "category is required");
+        require(bytes(_date).length > 0, "date is required");
+
+        uint256 id = tracker[msg.sender].counter;
```
[the remark this answers is not in the document]
Comment (Arpaninformationtech) [reply backend/contracts/ExpenseTracker.sol:67]: ok

File: backend/contracts/ExpenseTracker.sol (hunk 1 above)
Comment (Arpaninformationtech) [reply backend/contracts/ExpenseTracker.sol:65]: ok

**Comment by Arpaninformationtech** [comment]
is it now ok @shikhar229169 ??

File: backend/contracts/ExpenseTracker.sol [hunk 4]
```diff
@@ -17,75 +18,91 @@ contract ExpenseTracker {
...
-        transactions[msg.sender][_id] = Transaction(
-            _id,
+        require(bytes(_category).length > 0, "category is required");
+        require(bytes(_date).length > 0, "date is required");
+        require(isValidDateFormat(_date), "invalid date format");
+        require(isValidDate(_date), "invalid date");
+
+        uint256 id = tracker[msg.sender].total;
+
+        Transaction memory expense = Transaction(
```
Comment (shikhar229169) [inline backend/contracts/ExpenseTracker.sol:42]: The transaction should be stored in transactions mapping
mapping(address => mapping(string => Transaction)) private transactions;

Comment (shikhar229169) [reply backend/contracts/ExpenseTracker.sol:42]: Actually I was saying that the Transaction of the user is assigned to expense which is local to function, but it should be added to the mapping.
transactions[msg.sender][id] = expense
Ans same thing for income.

Comment (Arpaninformationtech) [reply backend/contracts/ExpenseTracker.sol:42]: transactions[msg.sender][id] = expense
And same thing for income.     These to be added where can you specify the line no.??

**Comment by J0SAL** [comment]
I realized the PR is been open for so long, but just wanted to update you that we can't merge it till the corresponding frontend is developed, either you can work on it (via diff issue) or we can wait for someone else. (you can create the issue for this too)

File: backend/contracts/ExpenseTracker.sol [hunk 5]
```diff
@@ -17,75 +18,91 @@ contract ExpenseTracker {
...
         string memory _date,
         string memory _category
     ) public {
-        require(all_ids[_id] == false, "invalid id");
-        transactions[msg.sender][_id] = Transaction(
-            _id,
+        require(bytes(_category).length > 0, "category is required");
+        require(bytes(_date).length > 0, "date is required");
+        require(isValidDateFormat(_date), "invalid date format");
+        require(isValidDate(_date), "invalid date");
```
[the remark this answers is not in the document]
Comment (Arpaninformationtech) [reply backend/contracts/ExpenseTracker.sol:38]: Ya I think we can but can't do it today I need 2 days at least plz tomorrow I have cllg exam🙂

**Comment by Arpaninformationtech** [comment]
> > @J0SAL can I create a new issue in 2 days?? And also what is diff issue and how to do it??
> 
> Yes you can create it wherever you want, for the other issue the idea would be to update the frontend to align with the smart contract

But If other contributora don't get to see the smart contract then how can they improve the front end according to it's requirements??

**Comment by J0SAL** [comment]
> > > @J0SAL can I create a new issue in 2 days?? And also what is diff issue and how to do it??
> > 
> > 
> > Yes you can create it wherever you want, for the other issue the idea would be to update the frontend to align with the smart contract
> 
> But If other contributora don't get to see the smart contract then how can they improve the front end according to it's requirements??

For that you can mention the exact requirements while creating an issue
